File: python/examples/hypercore_blocks.py
#!/usr/bin/env python3
"""
HyperCore Block Data Example

Shows how to get real-time trades, orders, and block data via the HyperCore API.

This is the alternative to Info methods (all_mids, l2_book, recent_trades) that
are not available on QuickNode endpoints.

Requirements:
    pip install hyperliquid-sdk

Usage:
    export ENDPOINT="https://your-endpoint.hype-mainnet.quiknode.pro/TOKEN"
    python hypercore_blocks.py
"""

import os
import sys

from hyperliquid_sdk import HyperliquidSDK

ENDPOINT = os.environ.get("ENDPOINT") or os.environ.get("QUICKNODE_ENDPOINT")

if not ENDPOINT:
    print("Error: Set ENDPOINT environment variable")
    print("  export ENDPOINT='https://your-endpoint.hype-mainnet.quiknode.pro/TOKEN'")
    sys.exit(1)


def main():
    print("=" * 50)
    print("HyperCore Block Data")
    print("=" * 50)

    # Single SDK instance - access everything through sdk.info(), sdk.core(), sdk.evm()
    sdk = HyperliquidSDK(ENDPOINT)
    hc = sdk.core()

    # Latest block number
    print("\n1. Latest Block:")
    block_num = hc.latest_block_number()
    print(f"   Block #{block_num}")

    # Recent trades
    print("\n2. Recent Trades (all coins):")
    trades = hc.latest_trades(count=5)
    for t in trades[:5]:
        side = "BUY" if t.get("side") == "B" else "SELL"
        print(f"   {side} {t.get('sz')} {t.get('coin')} @ ${t.get('px')}")

    # Recent BTC trades only
    print("\n3. BTC Trades:")
    btc_trades = hc.latest_trades(count=10, coin="BTC")
    for t in btc_trades[:3]:
        side = "BUY" if t.get("side") == "B" else "SELL"
        print(f"   {side} {t.get('sz')} @ ${t.get('px')}")
    if len(btc_trades) == 0:
        print("   No BTC trades in recent blocks")

    # Get a specific block
    print("\n4. Get Block Data:")
    block = hc.get_block(block_num - 1)
    if block:
        print(f"   Block #{block_num - 1}")
        print(f"   Time: {block.get('block_time', 'N/A')}")
        print(f"   Events: {len(block.get('events', []))}")

    # Get batch of blocks
    print("\n5. Batch Blocks:")
    blocks = hc.get_batch_blocks(block_num - 5, block_num - 1)
    block_list = blocks.get("blocks", []) if isinstance(blocks, dict) else blocks
    print(f"   Retrieved {len(block_list)} blocks")

    print()
    print("=" * 50)
    print("Done!")


if __name__ == "__main__":
    main()

File: typescript/examples/hypercore_blocks.ts
#!/usr/bin/env npx ts-node
// @ts-nocheck
/**
 * HyperCore Block Data Example
 *
 * Shows how to get real-time trades, orders, and block data via the HyperCore API.
 *
 * This is the alternative to Info methods (allMids, l2Book, recentTrades) that
 * are not available on QuickNode endpoints.
 *
 * Usage:
 *     export QUICKNODE_ENDPOINT="https://your-endpoint.hype-mainnet.quiknode.pro/YOUR_TOKEN"
 *     npx ts-node hypercore_blocks.ts
 */

import { HyperliquidSDK } from '@quicknode/hyperliquid-sdk';

const ENDPOINT = process.env.QUICKNODE_ENDPOINT;

if (!ENDPOINT) {
  console.error("Set QUICKNODE_ENDPOINT environment variable");
  process.exit(1);
}

async function main() {
  // Single SDK instance — access everything through sdk.info, sdk.core, sdk.evm, etc.
  const sdk = new HyperliquidSDK(ENDPOINT);
  const hc = sdk.core;

  console.log("=".repeat(50));
  console.log("HyperCore Block Data");
  console.log("=".repeat(50));

  // Latest block number
  console.log("\n1. Latest Block:");
  const blockNum = await hc.latestBlockNumber();
  console.log(`   Block #${blockNum}`);

  // Recent trades
  console.log("\n2. Recent Trades (all coins):");
  const trades = await hc.latestTrades(5);
  for (const t of trades.slice(0, 5)) {
    const side = t.side === "B" ? "BUY" : "SELL";
    console.log(`   ${side} ${t.sz} ${t.coin} @ $${t.px}`);
  }

  // Recent BTC trades only
  console.log("\n3. BTC Trades:");
  const btcTrades = await hc.latestTrades(10, "BTC");
  for (const t of btcTrades.slice(0, 3)) {
    const side = t.side === "B" ? "BUY" : "SELL";
    console.log(`   ${side} ${t.sz} @ $${t.px}`);
  }
  if (btcTrades.length === 0) {
    console.log("   No BTC trades in recent blocks");
  }

  // Get a specific block
  console.log("\n4. Get Block Data:");
  const block = await hc.getBlock(blockNum - 1);
  console.log(`   Block #${blockNum - 1}`);
  console.log(`   Time: ${block.block_time || 'N/A'}`);
  console.log(`   Events: ${(block.events || []).length}`);

  // Get batch of blocks
  console.log("\n5. Batch Blocks:");
  const blocks = await hc.getBatchBlocks(blockNum - 5, blockNum - 1);
  console.log(`   Retrieved ${(blocks.blocks || []).length} blocks`);

  console.log("\n" + "=".repeat(50));
  console.log("Done!");
}

main().catch(console.error);
